crates/pcb-zen-core/src/graph/starlark.rs

```diff
@@ -0,0 +1,487 @@
+    fn wrap_matcher_error<'v>(
+        &self,
+        original_error: starlark::Error,
+        eval: &mut Evaluator<'v, '_, '_>,
+    ) -> starlark::Error
+    where
+        V: ValueLike<'v>,
+    {
+        let Some(cs) = eval.call_stack_top_location() else {
+            return original_error;
+        };
+
+        let Some(path_val) = self
+            .path_value
+            .to_value()
+            .downcast_ref::<PathValueGen<Value>>()
+        else {
+            return original_error;
+        };
+
+        // Build child diagnostic directly from the Starlark error
+        let child = crate::diagnostics::Diagnostic::from(original_error);
+
+        // Create parent diagnostic with path.matches() context
+        let parent = crate::diagnostics::Diagnostic {
+            path: cs.file.filename().to_string(),
+            span: Some(cs.resolve_span()),
+            severity: starlark::errors::EvalSeverity::Error,
+            body: format!("{} failed to match sequence", path_val.description()),
+            call_stack: None,
+            child: Some(Box::new(child)),
+            source_error: None,
+        };
+
+        parent.into()
+    }
+}
+
+// Implementations for ModuleGraphValue
+impl<V: ValueLifetimeless> std::fmt::Display for ModuleGraphValueGen<V> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ModuleGraph")
+    }
+}
+
+#[starlark_value(type = "ModuleGraph")]
+impl<'v, V: ValueLike<'v>> StarlarkValue<'v> for ModuleGraphValueGen<V>
+where
+    Self: ProvidesStaticType<'v>,
+{
+    fn get_attr(&self, attr: &str, heap: &'v Heap) -> Option<Value<'v>> {
+        match attr {
+            "paths" => {
+                let callable = PathsCallableGen {
+                    module: self.module.to_value(),
+                    graph: self.graph.clone(),
+                };
+                Some(heap.alloc_complex(callable))
+            }
+            _ => None,
+        }
+    }
+}
+
+// PathsCallable implementation
+impl<V: ValueLifetimeless> std::fmt::Display for PathsCallableGen<V> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "paths")
+    }
+}
+
+#[starlark_value(type = "builtin_function_or_method")]
+impl<'v, V: ValueLike<'v>> StarlarkValue<'v> for PathsCallableGen<V>
+where
+    Self: ProvidesStaticType<'v>,
+{
+    fn invoke(
+        &self,
+        _me: Value<'v>,
+        args: &Arguments<'v, '_>,
+        eval: &mut Evaluator<'v, '_, '_>,
+    ) -> starlark::Result<Value<'v>> {
+        let heap = eval.heap();
+
+        // Extract arguments from named parameters
+        let args_map = args.names_map()?;
+        let start = args_map.get(&heap.alloc_str("start")).copied();
+        let end = args_map.get(&heap.alloc_str("end")).copied();
+        let max_depth = args_map
+            .get(&heap.alloc_str("max_depth"))
+            .and_then(|v| v.unpack_i32())
+            .unwrap_or(10) as usize;
```

Comment:
do we detect cycles in the graph pathfinding algorithm? if so, wdyt about making the max_depth opt-in but by default leaving it uncapped? I'm a little worried about a situation where we might have a handful of components in series and we quickly blow through this limit and it won't be obvious why we can't find a path

Reply:
yes, we avoid cycles by tracking visited edges, vertices. will make max_depth opt in